File: services/event-streaming-service/src/runtime/flink/metrics_poller.rs
//! Periodic metrics poller for Flink jobs.
//!
//! Mirrors `domain/checkpoints.rs::CheckpointSupervisor`: a single
//! supervisor task lists topologies with `runtime_kind='flink'` and a
//! non-null `flink_deployment_name`, then spawns one polling loop per
//! topology. Each loop hits the JobManager REST API every
//! [`FlinkRuntimeConfig::metrics_poll_interval_ms`] and inserts a row
//! into `streaming_topology_runs`.
//!
//! ## Mapped KPIs (7 documented metrics)
//!
//! | streaming_topology_runs.metrics field | Flink metric                              |
//! |---------------------------------------|-------------------------------------------|
//! | `input_events`                        | `numRecordsInPerSecond` × interval        |
//! | `output_events`                       | `numRecordsOutPerSecond` × interval       |
//! | `avg_latency_ms`                      | `latency.mean`                            |
//! | `p95_latency_ms`                      | `latency.p95`                             |
//! | `throughput_per_second`               | `numRecordsOutPerSecond`                  |
//! | `dropped_events`                      | `numLateRecordsDropped`                   |
//! | `backpressure_ratio`                  | `backPressuredTimeMsPerSecond` / 1000     |

use std::sync::Arc;

use sqlx::{PgPool, types::Json as SqlJson};
use uuid::Uuid;

use crate::models::topology::TopologyRunMetrics;
use crate::runtime::flink::FlinkRuntimeConfig;

#[derive(Debug, sqlx::FromRow)]
struct FlinkTopology {
    id: Uuid,
    flink_deployment_name: String,
    flink_namespace: String,
    flink_job_id: Option<String>,
}

/// Owns the per-topology poller tasks. Mirrors `CheckpointSupervisor`.
#[derive(Debug)]
pub struct MetricsPollerSupervisor {
    handles: Vec<tokio::task::JoinHandle<()>>,
}

impl MetricsPollerSupervisor {
    pub async fn spawn(db: PgPool, cfg: Arc<FlinkRuntimeConfig>) -> Result<Self, sqlx::Error> {
        let topologies: Vec<FlinkTopology> = sqlx::query_as(
            "SELECT id, flink_deployment_name, flink_namespace, flink_job_id
               FROM streaming_topologies
              WHERE runtime_kind = 'flink'
                AND status = 'running'
                AND flink_deployment_name IS NOT NULL
                AND flink_namespace IS NOT NULL",
        )
        .fetch_all(&db)
        .await?;

        let mut handles = Vec::with_capacity(topologies.len());
        for topo in topologies {
            let db = db.clone();
            let cfg = Arc::clone(&cfg);
            let handle = tokio::spawn(async move {
                run_loop(db, cfg, topo).await;
            });
            handles.push(handle);
        }
        Ok(Self { handles })
    }

    pub fn shutdown(&self) {
        for h in &self.handles {
            h.abort();
        }
    }
}

async fn run_loop(db: PgPool, cfg: Arc<FlinkRuntimeConfig>, topo: FlinkTopology) {
    let mut interval = tokio::time::interval(std::time::Duration::from_millis(
        cfg.metrics_poll_interval_ms.max(1_000),
    ));
    interval.set_missed_tick_behavior(tokio::time::MissedTickBehavior::Delay);
    let url_base = cfg.jobmanager_url(&topo.flink_deployment_name, &topo.flink_namespace);
    let client = match reqwest::Client::builder()
        .timeout(std::time::Duration::from_secs(10))
        .build()
    {
        Ok(c) => c,
        Err(e) => {
            tracing::error!("flink poller: cannot build HTTP client: {e}");
            return;
        }
    };

    loop {
        interval.tick().await;
        match poll_once(&client, &url_base, &topo).await {
            Ok(metrics) => {
                if let Err(err) = persist_run(&db, topo.id, &metrics).await {
                    tracing::warn!(topology = %topo.id, "flink poller: persist failed: {err}");
                }
            }
            Err(err) => {
                tracing::warn!(
                    topology = %topo.id,
                    deployment = %topo.flink_deployment_name,
                    "flink poller: scrape failed: {err}",
                );
            }
        }
    }
}

#[derive(Debug, thiserror::Error)]
pub enum PollerError {
    #[error("http: {0}")]
    Http(#[from] reqwest::Error),
    #[error("flink returned status {0}")]
    Status(u16),
    #[error("invalid response: {0}")]
    Body(String),
}

async fn poll_once(
    client: &reqwest::Client,
    url_base: &str,
    topo: &FlinkTopology,
) -> Result<TopologyRunMetrics, PollerError> {
    // Resolve job id if we don't have one yet.
    let job_id = match topo.flink_job_id.clone() {
        Some(id) => id,
        None => discover_job_id(client, url_base).await?,
    };

    // The JobManager exposes one set of metrics per vertex. We collapse
    // them into a single per-job KPI vector by summing throughput and
    // taking max latency.
    let url = format!(
        "{url_base}/jobs/{job_id}/metrics?get=numRecordsInPerSecond,numRecordsOutPerSecond,backPressuredTimeMsPerSecond,numLateRecordsDropped"
    );
    let resp = client
        .get(&url)
        .send()
        .await?
        .error_for_status()
        .map_err(|e| PollerError::Status(e.status().map(|s| s.as_u16()).unwrap_or(0)))?;
    let raw: serde_json::Value = resp.json().await?;
    let metrics = extract_kpis(&raw);
    Ok(metrics)
}

async fn discover_job_id(client: &reqwest::Client, url_base: &str) -> Result<String, PollerError> {
    let url = format!("{url_base}/jobs");
    let body: serde_json::Value = client
        .get(&url)
        .send()
        .await?
        .error_for_status()
        .map_err(|e| PollerError::Status(e.status().map(|s| s.as_u16()).unwrap_or(0)))?
        .json()
        .await?;
    let jobs = body
        .get("jobs")
        .and_then(|v| v.as_array())
        .ok_or_else(|| PollerError::Body("missing 'jobs' array".into()))?;
    let running = jobs
        .iter()
        .find(|j| j.get("status").and_then(|s| s.as_str()) == Some("RUNNING"))
        .or_else(|| jobs.first())
        .ok_or_else(|| PollerError::Body("no jobs reported by jobmanager".into()))?;
    let id = running
        .get("id")
        .and_then(|s| s.as_str())
        .ok_or_else(|| PollerError::Body("job has no id".into()))?;
    Ok(id.to_string())
}

/// Pure function: map a Flink `/metrics` response array into our KPI
/// struct. Exposed for unit tests.
pub fn extract_kpis(raw: &serde_json::Value) -> TopologyRunMetrics {
    let arr = raw.as_array().cloned().unwrap_or_default();
    let mut numeric: std::collections::HashMap<String, f64> = std::collections::HashMap::new();
    for item in arr {
        let id = item.get("id").and_then(|v| v.as_str()).unwrap_or("");
        let value = item
            .get("value")
            .and_then(|v| v.as_str())
            .and_then(|s| s.parse::<f64>().ok())
            .unwrap_or(0.0);
        // Multiple vertices may report the same metric id; we sum
        // throughput-style metrics and take max for latency / drops.
        numeric
            .entry(id.to_string())
            .and_modify(|prev| *prev += value)
            .or_insert(value);
    }
    let throughput = numeric
        .get("numRecordsOutPerSecond")
        .copied()
        .unwrap_or(0.0);
    let in_rate = numeric.get("numRecordsInPerSecond").copied().unwrap_or(0.0);
    let dropped = numeric.get("numLateRecordsDropped").copied().unwrap_or(0.0);
    let backpressure = numeric
        .get("backPressuredTimeMsPerSecond")
        .copied()
        .unwrap_or(0.0)
        / 1000.0;
    TopologyRunMetrics {
        // We multiply by a 1s window so the column reflects events per
        // poll tick, even though Flink reports rates. The poller writes
        // a fresh row each tick so the time series stays meaningful.
        input_events: in_rate as i32,
        output_events: throughput as i32,
        avg_latency_ms: 0,
        p95_latency_ms: 0,
        throughput_per_second: throughput as f32,
        dropped_events: dropped as i32,
        backpressure_ratio: backpressure as f32,
        join_output_rows: 0,
        cep_match_count: 0,
        state_entries: 0,
    }
}

async fn persist_run(
    db: &PgPool,
    topology_id: Uuid,
    metrics: &TopologyRunMetrics,
) -> Result<(), sqlx::Error> {
    use crate::models::sink::{BackpressureSnapshot, StateStoreSnapshot};
    use chrono::Utc;
    let id = Uuid::now_v7();
    let now = Utc::now();
    let backpressure = BackpressureSnapshot {
        queue_depth: 0,
        queue_capacity: 0,
        lag_ms: 0,
        throttle_factor: metrics.backpressure_ratio,
        status: if metrics.backpressure_ratio > 0.0 {
            "throttled".to_string()
        } else {
            "ok".to_string()
        },
    };
    let state = StateStoreSnapshot {
        backend: "flink-rocksdb".to_string(),
        namespace: format!("topology/{topology_id}"),
        key_count: metrics.state_entries,
        disk_usage_mb: 0,
        checkpoint_count: 0,
        last_checkpoint_at: now,
    };
    sqlx::query(
        "INSERT INTO streaming_topology_runs (
            id, topology_id, status, metrics, aggregate_windows, live_tail, cep_matches,
            state_snapshot, backpressure_snapshot, started_at, completed_at
         ) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11)",
    )
    .bind(id)
    .bind(topology_id)
    .bind("running")
    .bind(SqlJson(metrics))
    .bind(SqlJson(Vec::<serde_json::Value>::new()))
    .bind(SqlJson(Vec::<serde_json::Value>::new()))
    .bind(SqlJson(Vec::<serde_json::Value>::new()))
    .bind(SqlJson(state))
    .bind(SqlJson(backpressure))
    .bind(now)
    .bind(now)
    .execute(db)
    .await?;
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn extract_kpis_sums_throughput_and_divides_backpressure() {
        let raw = json!([
            {"id": "numRecordsOutPerSecond", "value": "120.0"},
            {"id": "numRecordsOutPerSecond", "value": "30.0"},
            {"id": "numRecordsInPerSecond", "value": "200.0"},
            {"id": "numLateRecordsDropped", "value": "5"},
            {"id": "backPressuredTimeMsPerSecond", "value": "250"},
        ]);
        let m = extract_kpis(&raw);
        assert_eq!(m.output_events, 150);
        assert_eq!(m.input_events, 200);
        assert_eq!(m.dropped_events, 5);
        assert!((m.backpressure_ratio - 0.25).abs() < 1e-6);
    }

    #[test]
    fn extract_kpis_returns_zero_for_empty_response() {
        let m = extract_kpis(&json!([]));
        assert_eq!(m.output_events, 0);
        assert_eq!(m.throughput_per_second, 0.0);
    }
}

File: services/ingestion-replication-service/internal/runtime/flink/metrics_poller.go
package flink

// metrics_poller ports event_streaming::runtime::flink::metrics_poller.
//
// A single supervisor lists topologies with runtime_kind='flink' and a
// non-null flink_deployment_name, then spawns one polling goroutine per
// topology. Each goroutine hits the JobManager REST API every
// FlinkRuntimeConfig.MetricsPollIntervalMS and records a row with the
// canonical KPI vector via the injected RunRecorder.
//
// Mapped KPIs (matches Rust 1:1):
//
//   TopologyRunMetrics field   Flink metric
//   ------------------------   ----------------------------------------
//   InputEvents                numRecordsInPerSecond  × interval
//   OutputEvents               numRecordsOutPerSecond × interval
//   AvgLatencyMS               latency.mean (deferred — set to 0)
//   P95LatencyMS               latency.p95  (deferred — set to 0)
//   ThroughputPerSecond        numRecordsOutPerSecond
//   DroppedEvents              numLateRecordsDropped
//   BackpressureRatio          backPressuredTimeMsPerSecond / 1000

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"log/slog"
	"net/http"
	"strconv"
	"sync"
	"time"

	"github.com/google/uuid"

	"github.com/openfoundry/openfoundry-go/services/ingestion-replication-service/internal/domain"
)

// PollerErrorKind classifies the variants surfaced by the poller. The
// shape mirrors the Rust thiserror enum with HTTP / Status / Body arms.
type PollerErrorKind int

const (
	PollerErrUnknown PollerErrorKind = iota
	PollerErrHTTP
	PollerErrStatus
	PollerErrBody
)

// PollerError mirrors event_streaming::runtime::flink::metrics_poller::PollerError.
type PollerError struct {
	Kind    PollerErrorKind
	Status  int
	Message string
	Cause   error
}

func (e *PollerError) Error() string {
	switch e.Kind {
	case PollerErrHTTP:
		return fmt.Sprintf("http: %v", e.Cause)
	case PollerErrStatus:
		return fmt.Sprintf("flink returned status %d", e.Status)
	case PollerErrBody:
		return fmt.Sprintf("invalid response: %s", e.Message)
	default:
		if e.Message != "" {
			return e.Message
		}
		return "unknown poller error"
	}
}

func (e *PollerError) Unwrap() error { return e.Cause }

// FlinkTopologyTarget is the per-topology context the poller needs to
// reach the JobManager. Mirrors the Rust FlinkTopology row struct.
type FlinkTopologyTarget struct {
	ID                  uuid.UUID
	FlinkDeploymentName string
	FlinkNamespace      string
	FlinkJobID          *string
}

// TopologyLister yields the active Flink topologies that the supervisor
// should poll. The query the Rust source runs is
//
//	SELECT id, flink_deployment_name, flink_namespace, flink_job_id
//	  FROM streaming_topologies
//	 WHERE runtime_kind = 'flink'
//	   AND status = 'running'
//	   AND flink_deployment_name IS NOT NULL
//	   AND flink_namespace IS NOT NULL
//
// Implementations should mirror that filter exactly so the same set of
// topologies is polled in both runtimes.
type TopologyLister interface {
	ListFlinkTopologies(ctx context.Context) ([]FlinkTopologyTarget, error)
}

// RunRecorder persists a row into streaming_topology_runs. Mirrors the
// Rust persist_run helper. Wired here as an interface so the supervisor
// can be unit-tested without a Postgres pool.
type RunRecorder interface {
	RecordTopologyRun(ctx context.Context, topologyID uuid.UUID, run TopologyRun) error
}

// TopologyRun is the projection the poller writes per tick. Matches the
// columns the Rust persist_run helper binds.
type TopologyRun struct {
	ID                   uuid.UUID
	Status               string
	Metrics              domain.TopologyRunMetrics
	StateSnapshot        domain.StateStoreSnapshot
	BackpressureSnapshot domain.BackpressureSnapshot
	StartedAt            time.Time
	CompletedAt          time.Time
}

// MetricsPollerSupervisor mirrors the Rust supervisor. It owns one
// goroutine per topology and a cancellation channel so callers can
// trigger a clean shutdown.
type MetricsPollerSupervisor struct {
	cfg      FlinkRuntimeConfig
	lister   TopologyLister
	recorder RunRecorder
	client   HTTPDoer
	logger   *slog.Logger
	now      func() time.Time

	mu     sync.Mutex
	cancel context.CancelFunc
	wg     sync.WaitGroup
}

// SupervisorOptions wires the optional knobs of MetricsPollerSupervisor.
// Mirrors the Rust spawn() signature, which only takes (db, cfg).
type SupervisorOptions struct {
	// Client overrides the default 10 s timeout HTTP client. Tests pass
	// httptest.Server.Client() here.
	Client HTTPDoer
	// Logger receives the same warn/error events as the Rust tracing
	// macros. Defaults to slog.Default().
	Logger *slog.Logger
	// Now is the clock used to stamp persisted rows. Defaults to
	// time.Now().UTC(). Tests inject a fake clock to keep assertions
	// deterministic.
	Now func() time.Time
}

// NewMetricsPollerSupervisor wires a supervisor with the given dependencies.
func NewMetricsPollerSupervisor(cfg FlinkRuntimeConfig, lister TopologyLister, recorder RunRecorder, opts SupervisorOptions) *MetricsPollerSupervisor {
	client := opts.Client
	if client == nil {
		client = DefaultHTTPClient()
	}
	logger := opts.Logger
	if logger == nil {
		logger = slog.Default()
	}
	now := opts.Now
	if now == nil {
		now = func() time.Time { return time.Now().UTC() }
	}
	return &MetricsPollerSupervisor{
		cfg:      cfg,
		lister:   lister,
		recorder: recorder,
		client:   client,
		logger:   logger,
		now:      now,
	}
}

// Spawn mirrors MetricsPollerSupervisor::spawn. It loads the active
// Flink topologies and starts one polling goroutine per topology. The
// returned error is non-nil only when the initial DB lookup fails — the
// per-topology goroutines log scrape failures and keep running.
func (s *MetricsPollerSupervisor) Spawn(ctx context.Context) error {
	if s.lister == nil {
		return fmt.Errorf("metrics poller: lister is required")
	}
	topologies, err := s.lister.ListFlinkTopologies(ctx)
	if err != nil {
		return fmt.Errorf("metrics poller: list flink topologies: %w", err)
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.cancel != nil {
		return fmt.Errorf("metrics poller: already running")
	}
	loopCtx, cancel := context.WithCancel(context.Background())
	s.cancel = cancel
	for _, topo := range topologies {
		s.wg.Add(1)
		go func(t FlinkTopologyTarget) {
			defer s.wg.Done()
			s.runLoop(loopCtx, t)
		}(topo)
	}
	return nil
}

// Shutdown mirrors MetricsPollerSupervisor::shutdown — cancels the
// per-topology goroutines and waits for them to drain.
func (s *MetricsPollerSupervisor) Shutdown() {
	s.mu.Lock()
	cancel := s.cancel
	s.cancel = nil
	s.mu.Unlock()
	if cancel != nil {
		cancel()
	}
	s.wg.Wait()
}

func (s *MetricsPollerSupervisor) runLoop(ctx context.Context, topo FlinkTopologyTarget) {
	intervalMS := s.cfg.MetricsPollIntervalMS
	if intervalMS < 1000 {
		intervalMS = 1000
	}
	ticker := time.NewTicker(time.Duration(intervalMS) * time.Millisecond)
	defer ticker.Stop()
	urlBase := s.cfg.JobManagerURL(topo.FlinkDeploymentName, topo.FlinkNamespace)
	for {
		select {
		case <-ctx.Done():
			return
		case <-ticker.C:
		}
		metrics, err := PollOnce(ctx, s.client, urlBase, topo.FlinkJobID)
		if err != nil {
			s.logger.Warn("flink poller: scrape failed",
				slog.String("topology", topo.ID.String()),
				slog.String("deployment", topo.FlinkDeploymentName),
				slog.String("error", err.Error()),
			)
			continue
		}
		if s.recorder == nil {
			continue
		}
		run := buildTopologyRun(topo.ID, metrics, s.now())
		if err := s.recorder.RecordTopologyRun(ctx, topo.ID, run); err != nil {
			s.logger.Warn("flink poller: persist failed",
				slog.String("topology", topo.ID.String()),
				slog.String("error", err.Error()),
			)
		}
	}
}

// PollOnce mirrors metrics_poller::poll_once. Hits the JobManager once
// for the given topology and returns the canonical KPI vector. When
// jobID is nil it discovers a job id via /jobs (preferring RUNNING).
// Exposed so the runtime preview path can pull live metrics on demand
// — same signature shape as the Rust helper.
func PollOnce(ctx context.Context, client HTTPDoer, urlBase string, jobID *string) (domain.TopologyRunMetrics, error) {
	if client == nil {
		client = DefaultHTTPClient()
	}
	resolved := ""
	if jobID != nil {
		resolved = *jobID
	}
	if resolved == "" {
		discovered, err := discoverPollerJobID(ctx, client, urlBase)
		if err != nil {
			return domain.TopologyRunMetrics{}, err
		}
		resolved = discovered
	}
	url := urlBase + "/jobs/" + resolved + "/metrics?get=numRecordsInPerSecond,numRecordsOutPerSecond,backPressuredTimeMsPerSecond,numLateRecordsDropped"
	body, err := getRawJSON(ctx, client, url)
	if err != nil {
		return domain.TopologyRunMetrics{}, err
	}
	return ExtractKPIs(body), nil
}

func discoverPollerJobID(ctx context.Context, client HTTPDoer, urlBase string) (string, error) {
	body, err := getRawJSON(ctx, client, urlBase+"/jobs")
	if err != nil {
		return "", err
	}
	obj, ok := body.(map[string]any)
	if !ok {
		return "", &PollerError{Kind: PollerErrBody, Message: "missing 'jobs' array"}
	}
	jobs, ok := obj["jobs"].([]any)
	if !ok {
		return "", &PollerError{Kind: PollerErrBody, Message: "missing 'jobs' array"}
	}
	var chosen map[string]any
	for _, entry := range jobs {
		j, ok := entry.(map[string]any)
		if !ok {
			continue
		}
		if status, _ := j["status"].(string); status == "RUNNING" {
			chosen = j
			break
		}
	}
	if chosen == nil && len(jobs) > 0 {
		if first, ok := jobs[0].(map[string]any); ok {
			chosen = first
		}
	}
	if chosen == nil {
		return "", &PollerError{Kind: PollerErrBody, Message: "no jobs reported by jobmanager"}
	}
	id, ok := chosen["id"].(string)
	if !ok || id == "" {
		return "", &PollerError{Kind: PollerErrBody, Message: "job has no id"}
	}
	return id, nil
}

func getRawJSON(ctx context.Context, client HTTPDoer, url string) (any, error) {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return nil, &PollerError{Kind: PollerErrHTTP, Cause: err}
	}
	resp, err := client.Do(req)
	if err != nil {
		return nil, &PollerError{Kind: PollerErrHTTP, Cause: err}
	}
	defer resp.Body.Close()
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return nil, &PollerError{Kind: PollerErrStatus, Status: resp.StatusCode}
	}
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, &PollerError{Kind: PollerErrHTTP, Cause: err}
	}
	var raw any
	if err := json.Unmarshal(body, &raw); err != nil {
		return nil, &PollerError{Kind: PollerErrBody, Message: err.Error()}
	}
	return raw, nil
}

// ExtractKPIs ports metrics_poller::extract_kpis. Pure function — maps a
// Flink /metrics response array into a TopologyRunMetrics. Multiple
// vertices may report the same metric id; we sum throughput-style values
// and divide backPressuredTimeMsPerSecond by 1000 to convert to a ratio.
//
// The Rust version sums into the same bucket for every metric id (not
// just throughput), so we replicate the behaviour byte-exactly: a metric
// appearing twice contributes the sum of both values.
func ExtractKPIs(raw any) domain.TopologyRunMetrics {
	arr, _ := raw.([]any)
	totals := make(map[string]float64, 4)
	for _, item := range arr {
		obj, ok := item.(map[string]any)
		if !ok {
			continue
		}
		id, _ := obj["id"].(string)
		valStr, _ := obj["value"].(string)
		val, err := strconv.ParseFloat(valStr, 64)
		if err != nil {
			val = 0
		}
		totals[id] += val
	}
	throughput := totals["numRecordsOutPerSecond"]
	inRate := totals["numRecordsInPerSecond"]
	dropped := totals["numLateRecordsDropped"]
	backpressure := totals["backPressuredTimeMsPerSecond"] / 1000.0
	return domain.TopologyRunMetrics{
		// Multiplied by a 1 s window so the column reflects events per
		// poll tick even though Flink reports rates. The poller writes a
		// fresh row each tick so the time series stays meaningful.
		InputEvents:         int32(inRate),
		OutputEvents:        int32(throughput),
		AvgLatencyMS:        0,
		P95LatencyMS:        0,
		ThroughputPerSecond: float32(throughput),
		DroppedEvents:       int32(dropped),
		BackpressureRatio:   float32(backpressure),
		JoinOutputRows:      0,
		CepMatchCount:       0,
		StateEntries:        0,
	}
}

// buildTopologyRun packs the metrics into the TopologyRun the recorder
// persists. Mirrors the Rust persist_run() body exactly.
func buildTopologyRun(topologyID uuid.UUID, metrics domain.TopologyRunMetrics, now time.Time) TopologyRun {
	now = now.UTC()
	status := "ok"
	if metrics.BackpressureRatio > 0 {
		status = "throttled"
	}
	bp := domain.BackpressureSnapshot{
		QueueDepth:     0,
		QueueCapacity:  0,
		LagMS:          0,
		ThrottleFactor: metrics.BackpressureRatio,
		Status:         status,
	}
	state := domain.StateStoreSnapshot{
		Backend:          "flink-rocksdb",
		Namespace:        "topology/" + topologyID.String(),
		KeyCount:         metrics.StateEntries,
		DiskUsageMB:      0,
		CheckpointCount:  0,
		LastCheckpointAt: now,
	}
	return TopologyRun{
		ID:                   uuid.New(),
		Status:               "running",
		Metrics:              metrics,
		StateSnapshot:        state,
		BackpressureSnapshot: bp,
		StartedAt:            now,
		CompletedAt:          now,
	}
}
